### ISSUE TITLE
Add built-in runtime profiling for data fetch, forward, and optimizer

### ISSUE THREAD

**Issue by mjenrungrot** [mjenrungrot/model-training-framework#13]
## Summary

Expose lightweight, opt-in profiling for four core timings **per dataloader**:

- `profile/train/dl_{loader_name}/time_data_ms`: time to fetch the next batch from specific dataloader
- `profile/train/dl_{loader_name}/time_forward_ms`: model forward (and loss) time for specific dataloader
- `profile/train/dl_{loader_name}/time_backward_ms`: backward pass time for specific dataloader
- `profile/train/dl_{loader_name}/time_optimizer_ms`: optimizer step time for specific dataloader

Use the existing `GenericTrainerConfig.profile_training` flag to enable this, with minimal overhead and clean integration via hooks. Provide consistent train/val metric names and DDP-safe logging.

## Motivation

- Make bottlenecks obvious during experimentation, especially with multiple dataloaders and AMP/scaler.
- Enable apples-to-apples comparisons across models/datasets without requiring user code changes.
- **Identify performance differences between different dataloaders in multi-dataloader setups.**
- Leverage existing hook system and `profile_training` config; avoid heavy PyTorch Profiler overhead.

## Scope

- Training loop (multi-dataloader) for both train and validation phases.
- **Per-dataloader granularity for all metrics.**
- Lightweight timers with optional CUDA sync for accuracy.
- No external profiler dependency.
- Out of scope: end-to-end flame graphs, kernel-level tracing, autograd graph profiling.

## Proposed Design

- Use new forward hooks around `training_step_fn` so we can time the true forward+loss region precisely.
- Use existing optimizer hooks to time the optimizer step.
- Use existing backward hooks to time the backward pass.
- **Track current dataloader name/index in trainer state to enable per-dataloader metrics.**
- For batch extraction, provide:
  - A no-core-change approximation: "batch wait" time between `on_train_batch_end` and `on_train_batch_start` (captured in a hook).
  - Optional exact timing: measure `next(iterator)` inside the multi-dataloader iterator and expose it back to the trainer when `profile_training=True`.

## API Changes

- Add two optional hook methods to `TrainerHooks`:
  - `on_before_forward(self, trainer)`
  - `on_after_forward(self, trainer, duration_ms: float)`
- Auto-register a new built-in `RuntimeProfilingHook` when `GenericTrainerConfig.profile_training` is `True`.
- **Add `current_dataloader_name` property to trainer to track active dataloader.**
- No breaking changes; default no-ops if not implemented.

## Metric Names

**Important**: Use `profile/` namespace with per-dataloader granularity to avoid conflicts with user metrics.

Train (per dataloader):
- `profile/train/dl_{loader_name}/time_data_ms`
- `profile/train/dl_{loader_name}/time_forward_ms`
- `profile/train/dl_{loader_name}/time_backward_ms`
- `profile/train/dl_{loader_name}/time_optimizer_ms`

Validation (per dataloader):
- `profile/val/dl_{loader_name}/time_data_ms`
- `profile/val/dl_{loader_name}/time_forward_ms`

Optional approximation:
- `profile/train/dl_{loader_name}/time_batch_wait_ms` (upper bound on dataloader wait, includes inter-batch work)

DDP: log only on primary rank.

## Where To Instrument

**Dataloader tracking:**
- File: `model_training_framework/trainer/core.py`: 
  - Add `self.current_dataloader_name` or `self.current_dataloader_idx` property
  - Update this property in `_train_epoch` and `_validation_epoch` loops when iterating over dataloaders

Forward timing and hook calls:
- File: `model_training_framework/trainer/core.py`: `_training_step(...)`
  - Before invoking `self.training_step_fn(...)`:
    - Call `hook_manager.call_hook("on_before_forward", self)`
    - Start timer; if CUDA, `torch.cuda.synchronize()`
  - After the call:
    - If CUDA, `torch.cuda.synchronize()`
    - Compute `duration_ms` and call `hook_manager.call_hook("on_after_forward", self, duration_ms)`
  - If `config.profile_training` and logger present, log `profile/train/dl_{loader_name}/time_forward_ms` (or `profile/val/dl_{loader_name}/...` when in validation).

Backward timing:
- File: `model_training_framework/trainer/core.py`: `_training_step(...)`
  - Use existing `on_before_backward` and `on_after_backward` hooks
  - Add timer around the backward pass region
  - Log `profile/train/dl_{loader_name}/time_backward_ms` when `profile_training=True`

Optimizer timing (use existing hooks):
- File: `model_training_framework/trainer/hooks.py`
  - Implement `RuntimeProfilingHook.on_before_optimizer_step`/`on_after_optimizer_step` to time and log `profile/*/dl_{loader_name}/time_optimizer_ms`.
  - The trainer already calls these hooks in `_optimizer_step(...)`.

Data fetch timing:
- Option A (exact):
  - File: `model_training_framework/trainer/multi_dataloader.py`: `_get_next_batch_from_loader(...)`
  - Surround `batch = next(iterator)` with timer and optional CUDA sync (usually CPU bound, CUDA sync optional/off).
  - Store `self.last_batch_fetch_ms` and `self.last_loader_name` on the iterator.
  - File: `model_training_framework/trainer/core.py`: `_train_epoch(...)`
    - Immediately after `for loader_idx, batch in iterator:` read `getattr(iterator, "last_batch_fetch_ms", None)` and `getattr(iterator, "last_loader_name", None)`, and if present and `profile_training`, log `profile/*/dl_{loader_name}/time_data_ms`.

- Option B (approximate, no core iterator changes):
  - File: `model_training_framework/trainer/hooks.py`
  - In `RuntimeProfilingHook`, time `on_train_batch_end` → `on_train_batch_start` and log `profile/*/dl_{loader_name}/time_batch_wait_ms` as an upper bound on dataloader wait.
  - This can be enabled even if Option A is deferred.

## Implementation Plan

1) Hooks API
- Add `on_before_forward` and `on_after_forward(self, trainer, duration_ms: float)` to `TrainerHooks`.
- File: `model_training_framework/trainer/hooks.py`

2) Dataloader tracking
- Add `current_dataloader_name` or similar property to trainer
- Update this property in training/validation loops
- File: `model_training_framework/trainer/core.py`

3) Forward timing in core
- In `_training_step(...)`, call new hooks and compute duration (with CUDA sync when available and profiling enabled).
- File: `model_training_framework/trainer/core.py`
- When `config.profile_training` is `True` and logger is present (and primary rank), log `profile/*/dl_{loader_name}/time_forward_ms`.

4) Backward timing in core
- In `_training_step(...)`, add timer around backward pass using existing hooks.
- Log `profile/train/dl_{loader_name}/time_backward_ms` when profiling is enabled.

5) Built-in profiling hook
- Add `RuntimeProfilingHook` class implementing:
  - Optimizer step timing via existing `on_before_optimizer_step`/`on_after_optimizer_step` to log `profile/*/dl_{loader_name}/time_optimizer_ms`.
  - Optional batch gap timing via `on_train_batch_end`/`on_train_batch_start` to log `profile/*/dl_{loader_name}/time_batch_wait_ms` (approximate).
  - Respect `trainer.config.logging.log_scalars_every_n_steps` (or an override).
  - Guard CUDA sync behind a flag.
- File: `model_training_framework/trainer/hooks.py`
- Auto-register when `GenericTrainerConfig.profile_training` is `True`.
- File: `model_training_framework/trainer/core.py` (post built-in hook registration)

6) Exact data fetch timing (optional but recommended)
- Instrument `MultiDataLoaderIterator._get_next_batch_from_loader(...)` to measure `next(iterator)` and store `self.last_batch_fetch_ms` and `self.last_loader_name`.
- File: `model_training_framework/trainer/multi_dataloader.py`
- In `_train_epoch(...)` and `_validation_epoch(...)`, after fetching a batch, read iterator attributes and log `profile/*/dl_{loader_name}/time_data_ms` when `profile_training` is `True`.
- File: `model_training_framework/trainer/core.py`

## Config and Usage

- Use existing `GenericTrainerConfig.profile_training` to toggle profiling.
- File: `model_training_framework/config/schemas.py`: `GenericTrainerConfig.profile_training` (already present)
- No new required flags; optionally expose a `HooksConfig` override to tweak hook behavior later.

## DDP and Logging

- Ensure all logs are gated by `ddp_is_primary(self.fabric)` and logger presence.
- Use `self.logger.log_metrics({...}, step=self.global_step)`.

## Validation Parity

- Mirror data/forward timings during validation (`profile/val/dl_{loader_name}/time_*`).
- Optimizer and backward timings remain train-only.

## Documentation

- Add a short section "Runtime Profiling" describing:
  - How to enable (`profile_training=True`)
  - Metric names and meaning (including per-dataloader structure)
  - Overheads and CUDA sync behavior
  - Train vs val metric prefixes
  - How to interpret per-dataloader metrics

## Test Plan

Unit tests with CPU-only environment:
- Enable `profile_training=True` on a toy model/loader; assert that metrics keys are logged at least once: `profile/train/dl_{loader_name}/time_forward_ms`, `profile/train/dl_{loader_name}/time_backward_ms`, `profile/train/dl_{loader_name}/time_optimizer_ms`, and either `profile/train/dl_{loader_name}/time_data_ms` (if Option A) or `profile/train/dl_{loader_name}/time_batch_wait_ms` (Option B).
- Verify zero/near-zero overhead when `profile_training=False` (e.g., hooks not registered; no metrics).

CUDA-available smoke test (skipped if CUDA not present):
- Ensure CUDA sync path works and logs metrics.

Validation loop:
- Assert `profile/val/dl_{loader_name}/time_forward_ms` and `profile/val/dl_{loader_name}/time_data_ms` are present during a validation epoch.

Multi-dataloader:
- With ≥2 loaders and `SamplingStrategy.WEIGHTED`, verify metrics are properly namespaced per dataloader.
- Verify metrics continue to log without duplication and without crashing when a loader exhausts.
- **Ensure each dataloader gets its own set of metrics with correct names.**

## Performance Considerations

- When disabled: zero impact aside from existence of new no-op hook methods.
- When enabled:
  - Forward timing adds two potential CUDA synchronizations per batch; allow disabling CUDA sync via a hook flag (default True for accuracy).
  - Backward timing adds two more CUDA syncs per batch.
  - Data fetch timing is CPU-bound; negligible overhead.
  - Optimizer timing adds optional CUDA syncs; acceptable for profiling runs.
  - Per-dataloader tracking adds minimal string formatting overhead.

## Backward Compatibility

- Hook API additions are optional; existing hooks remain valid.
- No change to trainer public API.
- Metrics logging is additive.

## Example Usage

Programmatic:
```python
GenericTrainerConfig(profile_training=True, ...)
```

Hydra/YAML-like:
```yaml
training:
  profile_training: true
```

With custom frequency:
- Provide a hook config override later (optional), or rely on `logging.log_scalars_every_n_steps`.

## Example Output

With multiple dataloaders (e.g., "imagenet", "cifar10"):
```
profile/train/dl_imagenet/time_data_ms: 12.5
profile/train/dl_imagenet/time_forward_ms: 45.2
profile/train/dl_imagenet/time_backward_ms: 38.1
profile/train/dl_imagenet/time_optimizer_ms: 15.3

profile/train/dl_cifar10/time_data_ms: 3.2
profile/train/dl_cifar10/time_forward_ms: 42.1
profile/train/dl_cifar10/time_backward_ms: 36.5
profile/train/dl_cifar10/time_optimizer_ms: 14.8
```

## Acceptance Criteria

With `profile_training=True`, logs include:
- Train: Per-dataloader `profile/train/dl_{loader_name}/time_forward_ms`, `profile/train/dl_{loader_name}/time_backward_ms`, and `profile/train/dl_{loader_name}/time_optimizer_ms`, plus either `profile/train/dl_{loader_name}/time_data_ms` (exact) or `profile/train/dl_{loader_name}/time_batch_wait_ms` (approx).
- Val: Per-dataloader `profile/val/dl_{loader_name}/time_forward_ms` and `profile/val/dl_{loader_name}/time_data_ms` (no optimizer/backward during val).
- Disabled by default; enabling does not break existing runs.
- DDP-safe: single-rank logging.
- Metrics use `profile/` namespace with per-dataloader granularity to avoid conflicts with user metrics.
- **Each dataloader in a multi-dataloader setup gets its own namespaced metrics.**

## Open Questions

- Should we also provide aggregated metrics across all dataloaders (e.g., `profile/train/total/time_forward_ms`)? Could be useful for overall performance monitoring.
- Should loader names be sanitized to ensure valid metric keys (e.g., replace spaces/special chars)?

## Future Work

- Add `time_scheduler_ms` per dataloader.
- Add percentile/rolling averages to logs.
- Optional integration with PyTorch Profiler for deeper dives (behind another flag).
- Aggregate statistics across dataloaders.
